docs/src/main/paradox/transactions.md

```diff
@@ -0,0 +1,72 @@
+
+To gracefully shutdown the stream and commit the current transaction you must call `shutdown()` on the `Control` materialized value to await all produced message acknowledgements and commit the final transaction.  
+
+### Simple Example
+
+Scala
+: @@ snip [transactionalSink](../../test/scala/sample/scaladsl/TransactionsExample.scala) { #transactionalSink }
+
+Java
+: @@ snip [transactionalSink](../../test/java/sample/javadsl/TransactionsExample.java) { #transactionalSink }
+
+### Recovery From Failure
+
+When any stage in the stream fails the whole stream will be torn down.  In the general case it's desirable to allow transient errors to fail the whole stream because they cannot be recovered from within the application.  Transient errors can be caused by network partitions, Kafka broker failures, `ProducerFencedException`'s from other application instances, and so on.  When the stream encounters transient errors then the current transaction will be aborted before the stream is torn down.  Any produced messages that were not committed will not be available to downstream consumers as long as those consumers are configured with `isolation.level = read_committed`.
+
+For transient errors we can choose to rely on the Kafka producer's configuration to retry, or we can handle it ourselves at the Akka Streams or Application layer.  Using the [`RestartSource`](https://doc.akka.io/docs/akka/current/stream/stream-error.html#delayed-restarts-with-a-backoff-stage) we can backoff connection attempts so that we don't hammer the Kafka cluster in a tight loop.
+
+Scala
+: @@ snip [transactionalFailureRetry](../../test/scala/sample/scaladsl/TransactionsExample.scala) { #transactionalFailureRetry }
+
+Java
+: @@ snip [transactionalFailureRetry](../../test/java/sample/javadsl/TransactionsExample.java) { #transactionalFailureRetry }
+
+## Caveats
+
+There are several scenarios that this library's implementation of Kafka transactions does not automatically account for.
+
+All of the scenarios covered in the @ref[At-Least-Once Delivery documentation](atleastonce.md) (Multiple Effects per Commit, Non-Sequential Processing, and Conditional Message Processing) are applicable to transactional scenarios as well.
+
+Only one application instance per `transactional.id` is allowed.  If two application instances with the same `transactional.id` are run at the same time then the instance that registers with Kafka's transaction coordinator second will throw a `ProducerFencedException` so it doesn't interfere with transactions in process by the first instance.  To distribute multiple transactional workflows for the same subscription the user must manually subdivide the subscription across multiple instances of the application.  This may be handled internally in future versions.
```

Comment:
In KafkaStreams they add a sufix with source topic name and partition number to transactional.id . It was the most challenging thing that I tried to explain in [comment](https://github.com/akka/reactive-kafka/issues/369#issuecomment-364974971) as in my opinion in most use cases there would be multiple "processors" that would handle the same work (same input topics). We cannot generate new transactional.id for new processing instance as then zombie fencing would not work either.

Reply:
This might still break exactly once processing. Simple example:
- single group name
- single topic T with 3 partitions: p1, p2, p3. Each partition has a lot of messages, p1 has p1m1,p1m2, ... p2 has p2m1,p2m2, etc. -  offset 0
- processing on 2 hosts/nodes/processes: A,B

After a system start (deployment/restart, etc.):
- node A registers, kafka assigns p1,p2,p3 partitions and the p1m1, p2m1, p3m1 messages are pulled.
- node B registers, kafka would reassign the partitions, for example A to p1,p2 and B to p3.
Process B could potentially pull the message from p3 - so it would process p3m1.

And now the race between A and B happens - the outcome should be that the processing of p3m1 on one of the nodes should finish successfully and the other should fail (InvalidProducerEpoch or ProducerFenced). 
In order to make this mechanism work both processes should use the same transactional.id - so during the partition attach and initTransaction the transactional.id epoch is bumped up. 

Because of that Transactional.id cannot be host related. 
See also https://www.confluent.io/blog/transactions-apache-kafka/ - but that does not give any solution... I cannot find out better solution than just coping the one from kafka streams...

This is _extremely_ important as when this bit is broken then we still have _at least once_ processing instead of _exactly once_.

Reply:
Please see also [my comment](https://github.com/akka/reactive-kafka/issues/369#issuecomment-380473894) here.

Reply:
Thanks a lot for joining this discussion @rafalmag , I was hoping you would provide your feedback!  Distributed transactional workflows are a hard problem to solve.  The implementation in KafkaStreams is complex.  I intentionally avoided this use case because I wanted to get the basic workflow for a single application instance working correctly.  I'm on the fence about implementing distributed transactional support in reactive-kafka because I think it can be handled by the client configuring appropriate consumer group subscriptions and the transaction.id themselves. This path loses out on the advantages of consumer group rebalancing, but does satisfy distributed transaction workflows.  Do you agree?

That said, I agree that it would be wonderful to have some way to support the distributed use case, so let's study our options.

In KafkaStreams the task's Producer's `transactional.id` is defined in [`StreamThread`](https://github.com/apache/kafka/blob/2e5d4af83f6b9f87fc84c0ff10462a5414ff8cb4/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L425).  It's a concatenation of the KafkaStreams config `application.id` and the `TaskId`. If you review [KIP-129](https://cwiki.apache.org/confluence/display/KAFKA/KIP-129%3A+Streams+Exactly-Once+Semantics) we see the design specifies one `KafkaProducer` per task.  So what is a task?

In Kafka Streams a task has an assigned set of topic partitions from the consumer group.  A task can contain multiple `StreamThread`'s to distribute work internally on the host machine.  A `PartitionGrouper` is used to assign tasks to partitions and the [`DefaultPartitionGrouper`](https://github.com/apache/kafka/blob/f26fbb9adcd66a31740da5c99f14a108bbc24304/streams/src/main/java/org/apache/kafka/streams/processor/DefaultPartitionGrouper.java) assigns each `TaskId` to a subset of topic partitions.  

How the `TaskId`'s are created is not clear to me, but it seems to be related to how many `Source`'s are available for the stream.  I think this may translate to the number of active consumers there are for a consumer group.  If that's the case then somehow Kafka Streams is aware of other consumers and assigns them ID's, so the assignment can happen, probably thru some metadata Kafka topic.

Obviously more research needs to be done here, but I agree that the simplest approach is what you already suggested by assigning a producer `transaction.id` per consumed topic partition.  I'll do a little more digging in the Kafka Streams codebase to see how `TaskId`'s are generated, but if it involves out of band communication to other app instances then I think the solution maybe too high level and complex for the reactive-kafka project.

Reply:
@seglo I really like the idea of `TransactionalMessage` in
```  
  def transactionalSource[K, V](settings: ConsumerSettings[K, V], subscription: Subscription): Source[TransactionalMessage[K, V], Control]
```
and
```
def transactionalSink[K, V](settings: ProducerSettings[K, V], transactionalId: String): 
Sink[Message[K, V, ConsumerMessage.PartitionOffset], Future[Done]] 
```

Rebalancing can be detected by consumer in source, transaction id should be set for producer in sink, but this approach really decouples source and sink. That is why I suggested in [comment](https://github.com/akka/reactive-kafka/issues/369#issuecomment-364974971) to implement something that receives the logic as a flow and does the creation of consumer, producer and necessary feedback loop in its internals, eg:
```
def withTransactions(consumerSettings: ConsumerSettings, producerSettings: ProducerSettings, flow: Flow[ConsumerRecord, Seq[ProducerRecord], Mat]): RunnableGraph[Mat]
```

Previously I thought that method could be implemented using:

```  
  def plainPartitionedSource[K, V](settings: ConsumerSettings[K, V], subscription: AutoSubscription): Source[(TopicPartition, Source[ConsumerRecord[K, V], NotUsed]), Control]
```	

so here we get a new substream for each topic-partition pair, so we can attach the flow and then some lazy producer (that would be created taking as an argument the TopicPartition). 

BTW: Maybe a corresponding method based on `TransactionalMessage` could be useful for that:
```
  def transactionalSource[K, V](settings: ConsumerSettings[K, V], subscription: AutoSubscription): Source[(TopicPartition, Source[TransactionalMessage[K, V], NotUsed]), Control]
```

In the sourcecode of `plainPartitionedSource` you will find a mechanism of emitting a new substream after rebalancing and a closing the old stream (in case of detach). 

Regarding:
> I'll also do a little more digging in the Kafka Streams codebase to see how TaskId's are generated, but if it involves out of band communication to other app instances then I think the solution maybe too high level and complex for the reactive-kafka project.

[TaskId](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java) in the Kafka Streams is just topicGorupId and partition in a form (`t_p`)

For setting the task Id no extra communication is required. The [org.apache.kafka.clients.consumer.Consumer](https://github.com/apache/kafka/blob/23d01c805bef7504abfa83ecac7e384d121a583a/clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java#L56) has a nice method:
```
public void subscribe(Collection<String> topics, ConsumerRebalanceListener callback);
```
and [ConsumerRebalanceListener](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.java) could be used to listen to the rebalancing events to know what partitions are revoked and what new are created.